flink-runtime/src/main/java/org/apache/flink/runtime/checkpoint/metadata/MetadataV2V3SerializerBase.java

```diff
@@ -415,7 +440,7 @@ KeyedStateHandle deserializeKeyedStateHandle(
             int baseSize = dis.readInt();
             List<KeyedStateHandle> base = new ArrayList<>(baseSize);
             for (int i = 0; i < baseSize; i++) {
-                base.add(deserializeKeyedStateHandle(dis, context));
+                base.add(Preconditions.checkNotNull(deserializeKeyedStateHandle(dis, context)));
```

Comment:
Why was this check aded?
`deserializeKeyedStateHandle` clearly can return null and is marked as `Nullable`.

Reply:
I see. Maybe it's better to skip nulls (i.e. don't add to the collection)?

I'm not sure whether it is guaranteed that there will be no nulls because this part comes from the nested backend (especially when switching changelog.enabled off => on).

Reply:
Since we already write the size of collection out, it's really strange that the materialized collection having `null` as its elements. I think a check here could act as a safety guard here.

Reply:
I agree, it's quite strange; but nevertheless possible. 
Currently, this is a normal case; with a null-check, this will be an exceptional case, recovery will fail in what was a normal case before.

Reply:
I prefer to throw exception here to act as a safety guard. Otherwise, if we take the slient exception here, we should at least add warnings.

Reply:
Maybe there is some misunderstanding, let me clarify what I mean:
1. `NULL_HANDLE = 0` - there is already an explicit constant that encodes null handles
2. I propose:
```
for (int i = 0; i < baseSize; i++) {
    KeyedStateHandle handle = deserializeKeyedStateHandle(dis, context);
    if (handle != null) {
        base.add(handle);
    }
}
```
If you think throwing exception or logging here is more appropriate I'm also fine with this approach.

Reply:
Thanks for your clarification, I think I did not misunderstand your ideas here. I just prefer strict limitation to avoid unexpcted user behavior. I will add a warning log based on your proposal.